File: the-lyceum-academy/src/lib/forum.ts
/**
 * Client for thelyceum.site/forum — the founder community. Signup happens
 * before any Lyceum account exists, so browse/apply calls are unauthenticated
 * (same reasoning as applications.py's submitApplication/getApplicationStatus).
 * Chat/connection calls carry the forum-issued token (stored in localStorage
 * after apply/login) since real user-to-user interaction needs more than a
 * client-supplied email to say who's posting.
 */
import { getApiBaseUrl } from './apiBase';

const API_BASE = getApiBaseUrl();
const TOKEN_KEY = 'lyceum_forum_token';

export function getForumToken(): string {
  return localStorage.getItem(TOKEN_KEY) || '';
}

export function setForumToken(token: string): void {
  localStorage.setItem(TOKEN_KEY, token);
}

export function clearForumToken(): void {
  localStorage.removeItem(TOKEN_KEY);
}

async function publicGetJson<T>(path: string): Promise<T> {
  const res = await fetch(`${API_BASE}${path}`);
  if (!res.ok) throw new Error((await res.json().catch(() => ({})))?.detail || `Request failed (${res.status})`);
  return res.json();
}

async function publicPostJson<T>(path: string, body: unknown): Promise<T> {
  const res = await fetch(`${API_BASE}${path}`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  });
  if (!res.ok) throw new Error((await res.json().catch(() => ({})))?.detail || `Request failed (${res.status})`);
  return res.json();
}

async function authedGetJson<T>(path: string): Promise<T> {
  const res = await fetch(`${API_BASE}${path}`, { headers: { 'X-Forum-Token': getForumToken() } });
  if (!res.ok) throw new Error((await res.json().catch(() => ({})))?.detail || `Request failed (${res.status})`);
  return res.json();
}

async function authedPostJson<T>(path: string, body: unknown): Promise<T> {
  const res = await fetch(`${API_BASE}${path}`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json', 'X-Forum-Token': getForumToken() },
    body: JSON.stringify(body),
  });
  if (!res.ok) throw new Error((await res.json().catch(() => ({})))?.detail || `Request failed (${res.status})`);
  return res.json();
}

export interface ForumGroup {
  id: string;
  name: string;
  kind: 'stage' | 'industry';
  description: string;
  is_active: number;
  member_count: number;
}

export function listForumGroups(): Promise<{ groups: ForumGroup[] }> {
  return publicGetJson('/forum/groups');
}

export type ForumRole = 'founder' | 'visitor';

export interface ForumApplyPayload {
  name: string;
  email: string;
  linkedin_url: string;
  role: ForumRole;
  business_email?: string;
  business_name?: string;
  industry?: string;
  stage?: string;
  photo_url?: string;
  looking_for_customers?: boolean;
  target_customer_description?: string;
}

export interface ForumMember {
  email: string;
  name: string;
  role: ForumRole;
  business_email: string;
  business_name: string;
  industry: string;
  stage: string;
  linkedin_url: string;
  photo_url: string;
  looking_for_customers: number;
  target_customer_description: string;
  referral_code: string;
  referred_by: string;
  joined_at: string;
  last_active_at: string;
  foundi_balance: number;
  groups: ForumGroup[];
}

export interface ForumApplyResult { ok: boolean; member: ForumMember; already_joined: boolean; forum_token: string; }

export async function applyToForum(payload: ForumApplyPayload): Promise<ForumApplyResult> {
  const res = await publicPostJson<ForumApplyResult>('/forum/apply', payload);
  setForumToken(res.forum_token);
  return res;
}

export async function forumLogin(email: string): Promise<{ member: ForumMember; forum_token: string }> {
  const res = await publicPostJson<{ member: ForumMember; forum_token: string }>('/forum/login', { email });
  setForumToken(res.forum_token);
  return res;
}

export function getForumMember(email: string): Promise<{ member: ForumMember }> {
  return publicGetJson(`/forum/me?email=${encodeURIComponent(email)}`);
}

// ── Suggested connections ───────────────────────────────────────────────────

export interface ForumConnection {
  id: string;
  member_email: string;
  match_email: string;
  kind: 'customer' | 'founder';
  reason: string;
  computed_at: string;
  match: ForumMember;
}

export function getForumConnections(): Promise<{ customers: ForumConnection[]; founders: ForumConnection[] }> {
  return authedGetJson('/forum/connections');
}

export function dismissForumConnection(id: string): Promise<{ ok: boolean }> {
  return authedPostJson(`/forum/connections/${encodeURIComponent(id)}/dismiss`, {});
}

// ── Chat — thread_id is 'group:<group_id>' or 'dm:<emailA>|<emailB>' (sorted) ─

export function dmThreadId(emailA: string, emailB: string): string {
  const [a, b] = [emailA.trim().toLowerCase(), emailB.trim().toLowerCase()].sort();
  return `dm:${a}|${b}`;
}

export interface ForumPitch { context: string; hypothesis: string; reasoning: string; risk: string; }

export interface ForumComment {
  id: string;
  message_id: string;
  author_email: string;
  verdict: 'good' | 'mixed' | 'bad';
  content: string;
  created_at: string;
}

export interface ForumMessage {
  id: string;
  thread_id: string;
  author_email: string;
  message_type: 'text' | 'pitch';
  body: string;
  pitch_context: string;
  pitch_hypothesis: string;
  pitch_reasoning: string;
  pitch_risk: string;
  created_at: string;
  comments: ForumComment[];
}

export function listForumMessages(threadId: string, since?: string): Promise<{ messages: ForumMessage[] }> {
  const qs = since ? `?since=${encodeURIComponent(since)}` : '';
  return authedGetJson(`/forum/threads/${encodeURIComponent(threadId)}/messages${qs}`);
}

export function postForumMessage(
  threadId: string, messageType: 'text' | 'pitch', body: string, pitch?: ForumPitch,
): Promise<{ ok: boolean; id: string; created_at: string }> {
  return authedPostJson(`/forum/threads/${encodeURIComponent(threadId)}/messages`, {
    message_type: messageType, body, pitch,
  });
}

export function postForumComment(
  messageId: string, verdict: 'good' | 'mixed' | 'bad', content: string,
): Promise<{ ok: boolean; id: string; created_at: string }> {
  return authedPostJson(`/forum/messages/${encodeURIComponent(messageId)}/comments`, { verdict, content });
}

File: backend/app/routers/forum.py
"""
Forum — public signup/browse/chat endpoints for thelyceum.site/forum. No
Firebase auth requirement: this is a pre-account signup funnel identified by
email, same style as applications.py/access_codes.py. Chat and connection
endpoints instead require a forum-issued token (X-Forum-Token header,
app.services.forum.issue_forum_token/verify_forum_token) since real
user-to-user interaction makes a bare client-supplied email spoofable.
Admin management (group CRUD, member rosters) lives in admin.py alongside
every other admin resource, guarded by admin.py's own `_auth`.
"""

from __future__ import annotations

from fastapi import APIRouter, Depends, Header, HTTPException, Request
from pydantic import BaseModel

from app.core.limiter import limiter
from app.services import forum as forum_svc

router = APIRouter(prefix="/forum", tags=["forum"])


def _forum_auth(x_forum_token: str = Header(...)) -> str:
    email = forum_svc.verify_forum_token(x_forum_token)
    if not email:
        raise HTTPException(status_code=401, detail="invalid_or_expired_forum_token")
    return email


@router.get("/groups")
def forum_groups():
    return {"groups": forum_svc.list_groups(active_only=True)}


class ForumApplyRequest(BaseModel):
    name: str
    email: str
    linkedin_url: str = ""
    role: str = "founder"
    business_email: str = ""
    business_name: str = ""
    industry: str = ""
    stage: str = ""
    photo_url: str = ""
    looking_for_customers: bool = False
    target_customer_description: str = ""


@router.post("/apply")
@limiter.limit("5/minute")
async def forum_apply(request: Request, req: ForumApplyRequest):
    result = await forum_svc.apply(
        name=req.name,
        email=req.email,
        linkedin_url=req.linkedin_url,
        role=req.role,
        business_email=req.business_email,
        business_name=req.business_name,
        industry=req.industry,
        stage=req.stage,
        photo_url=req.photo_url,
        looking_for_customers=req.looking_for_customers,
        target_customer_description=req.target_customer_description,
    )
    if not result.get("ok"):
        status = 402 if result.get("error") == "paid_plan_required" else 400
        raise HTTPException(status_code=status, detail=result.get("error"))
    return result


class ForumLoginRequest(BaseModel):
    email: str


@router.post("/login")
@limiter.limit("10/minute")
def forum_login(request: Request, req: ForumLoginRequest):
    result = forum_svc.login(req.email)
    if not result:
        raise HTTPException(status_code=404, detail="not_found")
    if result.get("error"):
        raise HTTPException(status_code=402, detail=result["error"])
    return result


@router.get("/me")
def forum_me(email: str):
    member = forum_svc.get_member(email)
    if not member:
        raise HTTPException(status_code=404, detail="not_found")
    forum_svc.touch_activity(member["email"])
    return {"member": member}


@router.get("/connections")
def forum_connections(email: str = Depends(_forum_auth)):
    return forum_svc.get_connections(email)


@router.post("/connections/{connection_id}/dismiss")
def forum_connections_dismiss(connection_id: str, email: str = Depends(_forum_auth)):
    result = forum_svc.dismiss_connection(connection_id, email)
    if not result.get("ok"):
        raise HTTPException(status_code=404, detail=result.get("error"))
    return result


@router.get("/threads/{thread_id}/messages")
def forum_thread_messages(thread_id: str, since: str | None = None, email: str = Depends(_forum_auth)):
    if not forum_svc.can_access_thread(thread_id, email):
        raise HTTPException(status_code=403, detail="not_a_participant")
    return {"messages": forum_svc.list_messages(thread_id, since)}


class ForumPitchIn(BaseModel):
    context: str = ""
    hypothesis: str = ""
    reasoning: str = ""
    risk: str = ""


class ForumPostMessageRequest(BaseModel):
    message_type: str
    body: str = ""
    pitch: ForumPitchIn | None = None


@router.post("/threads/{thread_id}/messages")
@limiter.limit("20/minute")
def forum_post_message(request: Request, thread_id: str, req: ForumPostMessageRequest, email: str = Depends(_forum_auth)):
    pitch = req.pitch or ForumPitchIn()
    result = forum_svc.post_message(
        thread_id=thread_id,
        author_email=email,
        message_type=req.message_type,
        body=req.body,
        pitch_context=pitch.context,
        pitch_hypothesis=pitch.hypothesis,
        pitch_reasoning=pitch.reasoning,
        pitch_risk=pitch.risk,
    )
    if not result.get("ok"):
        raise HTTPException(status_code=400, detail=result.get("error"))
    return result


class ForumCommentRequest(BaseModel):
    verdict: str
    content: str = ""


@router.post("/messages/{message_id}/comments")
@limiter.limit("30/minute")
def forum_post_comment(request: Request, message_id: str, req: ForumCommentRequest, email: str = Depends(_forum_auth)):
    result = forum_svc.add_comment(message_id, email, req.verdict, req.content)
    if not result.get("ok"):
        raise HTTPException(status_code=400, detail=result.get("error"))
    return result
